app implementation

silver-volt4 opened the pull request.

silver-volt4 committed:
wip push

silver-volt4 committed:
push

silver-volt4 committed:
what

silver-volt4 committed:
IT WORKS

silver-volt4 committed:
more stuff that works

silver-volt4 committed:
omg i love this

silver-volt4 committed:
read self as friend

silver-volt4 committed:
more options

silver-volt4 committed:
mac matching

silver-volt4 committed:
safer file reading

silver-volt4 committed:
improve picker gui

silver-volt4 committed:
paging, input buffering

silver-volt4 committed:
remove testing

silver-volt4 committed:
file writing

silver-volt4 committed:
move files

Co-authored-by: Charlotte Herngreen <CenTdemeern1@users.noreply.github.com>

silver-volt4 committed:
We have a working migration tool!

silver-volt4 committed:
fmt

silver-volt4 committed:
clippy

silver-volt4 committed:
updates

silver-volt4 committed:
use struct for extdata

silver-volt4 committed:
optimize reading phase

silver-volt4 committed:
updates

silver-volt4 committed:
started work on GUI

silver-volt4 committed:
text draw helper

silver-volt4 committed:
GUI!!!!!

silver-volt4 committed:
fmt

silver-volt4 committed:
cleanup

silver-volt4 committed:
fmt

silver-volt4 committed:
use utility functions

silver-volt4 committed:
cleanup

CenTdemeern1 commented on src/gui.rs:
```diff
@@ -0,0 +1,163 @@
+use ctru_sys as ctru;
+
+pub struct GUI {
+    pub screen: *mut C3D_RenderTarget,
```
This probably shouldn't be public
Also, I'm somewhat concerned about the memory safety of this; this kind of acts as "unsafe interior mutability"
What I'd do is make this private, and then make all the functions that draw to the screen require `&mut self` or something, so they need exclusive access and we can't get race conditions or data races or something. 
This is mostly just formalities though; the 3DS's threading model would make this kinda stuff difficult to do wrongly, *and* you're using syncdraw, but if possible it'd be nice to ensure everything is represented at least somewhat safely on a type level just for ease of mind and proper practice

CenTdemeern1 replied:
Will be dealt with through #2

CenTdemeern1 commented on src/gui.rs:
```diff
@@ -0,0 +1,163 @@
+    pub red: u32,
+    pub blue: u32,
+    pub dialog_overlay: u32,
+    pub highlight_color: u32,
```
These should be constants, they don't really change ever as far as I can tell.

CenTdemeern1 commented on src/gui.rs:
```diff
@@ -0,0 +1,163 @@
+        self._draw_text(text, x, y, flags | C2D_WithColor, self.red, scale);
+    }
+
+    fn _draw_text(&self, text: &C2D_Text, x: f32, y: f32, flags: u8, color: u32, scale: f32) {
```
This isn't Python, you don't need to prefix the name with an underscore; just making the function private is enough

silver-volt4 committed:
a

silver-volt4 committed:
*sigh*

silver-volt4 commented on src/phases/mapping.rs:
```diff
@@ -0,0 +1,413 @@
+    pub fn paint_whole_friend_picker(&self, pid: u32, friends: &MiiMap) {
+        const PAGE_SIZE: usize = 10;
+
+        unsafe {
```
remove this unsafe 
(Thanks Mike!)

silver-volt4 committed:
chore: add a readme file

silver-volt4 committed:
fix: incorrect markdown syntax

VS code has bamboozled me

silver-volt4 committed:
chore: GUI -> Gui

silver-volt4 committed:
feat: use new generic list picker for region selection

silver-volt4 committed:
feat: use single textbuffer instance

silver-volt4 committed:
feat: use new list picker everywhere

silver-volt4 committed:
refactor: wrap all logic in Scene structs

silver-volt4 committed:
chore: clippy fix

silver-volt4 committed:
chore: fix comment

silver-volt4 committed:
chore: clippy fixes

silver-volt4 committed:
refactor: split gui.rs into gui module

silver-volt4 committed:
chore: update packages

silver-volt4 committed:
feat: use ControlFlow to implement phase breaking instead of Result

CenTdemeern1 committed:
chore: cargo fmt

CenTdemeern1 committed:
feat: AbortMigration struct and control_flow module

CenTdemeern1 committed:
chore: merge and sort imports

CenTdemeern1 commented on src/friend_list.rs:
```diff
@@ -1,65 +1,40 @@
-        let mut friend_keys: [FriendKey; 100] = mem::transmute([1u8; 1600]);
-        let mut friend_info: [FriendInfo; 100] = mem::transmute([1u8; 22400]); // mem::zeroed()
+        let mut friend_keys: [FriendKey; 100] = mem::zeroed();
+        let mut friend_info: [FriendInfo; 100] = mem::zeroed();
```
These should be using the `FRIEND_LIST_SIZE` constant

CenTdemeern1 commented on src/friend_list.rs:
```diff
@@ -69,15 +44,14 @@ unsafe fn get_friend_info(handle: Handle) -> ([FriendKey; 100], [FriendInfo; 100
         *cmdbuf = 0x1A00C4;
         *cmdbuf.wrapping_add(1) = num_friends;
-        *cmdbuf.wrapping_add(2) = 0;
+        *cmdbuf.wrapping_add(2) = 1; // Mask non-ascii characters
```
Reminder: you said you should remove this

CenTdemeern1 commented on Cargo.toml:
```diff
@@ -7,6 +7,11 @@ edition = "2024"
 romfs_dir = "romfs"
+
+[build-dependencies]
+bindgen = "0.71.0"
```
This is no longer used

CenTdemeern1 commented on src/friend_list.rs:
```diff
@@ -89,27 +63,53 @@ unsafe fn get_friend_info(handle: Handle) -> ([FriendKey; 100], [FriendInfo; 100
+        for i in 0..num_friends {
+            let pid: u32 = friend_keys[i as usize].principalId;
+            let mii_bytes: [u8; 0x5C] =
+                // There seems to be a bug in how FriendInfo is deserialized.
+                // The Mii data is simply incorrect, the version should always be 3, but it isn't
```
Should be reported then right?

CenTdemeern1 replied:
Bump: should we report this bug?

silver-volt4 replied:
Probably yeah. I have no idea what to put in the issue though

And I don't want to wait for it to be fixed tbh

CenTdemeern1 replied:
We don't have to wait for it to be fixed, just reporting it is fine, we can continue on without a fix right

silver-volt4 replied:
Yeah we don't, the transmute is ugly but works

CenTdemeern1 committed:
refactor: panic_if_failed

CenTdemeern1 committed:
fix: hotfix for some mistakes i made during that refactor

CenTdemeern1 commented on src/extdata.rs:
```diff
@@ -1,167 +1,190 @@
-            ));
-            entry
+impl ExtdataArchive {
+    pub fn open(region: SwapdoodleRegion) -> Result<ExtdataArchive, ()> {
```
This should be changed to return either `Result<ExtdataArchive, ctru_sys::Result>` or `Option<ExtdataArchive>`
If you want to display the error codes if no archives are available, you should probably do the former

CenTdemeern1 replied:
Hm, or maybe even `Result<ExtdataArchive, NonZero<ctru_sys::Result>>`?

CenTdemeern1 commented on src/extdata.rs:
```diff
@@ -1,167 +1,190 @@
+
+            match R_SUCCEEDED(FSUSER_OpenArchive(
+                &mut extdata_handle as *mut _,
+                0x00000006, // ARCHIVE_EXTDATA
```
Why not just use `ARCHIVE_EXTDATA`

silver-volt4 replied:
Oh, that's a constant that exists. Neat.

silver-volt4 committed:
chore: use literal zeroes for simple data type initializers

CenTdemeern1 committed:
style: use inferred types

CenTdemeern1 commented on src/friend_list.rs:
```diff
@@ -1,115 +1,105 @@
+                mii[idx] = v;
+                idx += 1;
+            }
+        }
```
```suggestion
        let mut mii = [0; _];
        copy_nonoverlapping(
            cmdbuf.wrapping_add(2) as *mut _,
            mii.as_mut_ptr(),
            mii.len(),
        );
```
if it works in regards to endianness and stuff, just doing a memcpy is probably better

CenTdemeern1 commented on src/extdata.rs:
```diff
@@ -1,167 +1,190 @@
+    pub fn read_file(&self, path: &str) -> Vec<u8> {
+        const BATCH_SIZE: u32 = 1024;
+
+        let mut file = Vec::<u8>::new();
```
```suggestion
        let mut file = vec![];
```

CenTdemeern1 commented on src/extdata.rs:
```diff
@@ -1,167 +1,190 @@
+        let mut file = Vec::<u8>::new();
+
+        let mut path: Vec<u16> = path.encode_utf16().collect();
+        path.push(0);
```
```suggestion
        let path: Vec<u16> = path.encode_utf16().chain([0]).collect();
```
This seems like it would cause fewer allocations
Since it can collect the whole thing, including the null byte, in one go, instead of possibly having to extend the capacity of the vec during that push
([Compiler Explorer seems to back me up on this](https://godbolt.org/z/o998f5cbj))

CenTdemeern1 commented on src/extdata.rs:
```diff
@@ -1,167 +1,190 @@
+                offset += read as u64;
+                for i in 0..read {
+                    file.push(buffer[i as usize]);
+                }
```
```suggestion
                file.extend_from_slice(&buffer[0..read as usize]);
```

CenTdemeern1 commented on src/extdata.rs:
```diff
@@ -1,167 +1,190 @@
+        unsafe {
+            let mut handle: Handle = 0;
+            let mut path: Vec<u16> = path.encode_utf16().collect();
+            path.push(0);
```
Ditto

CenTdemeern1 commented on src/extdata.rs:
```diff
@@ -1,167 +1,190 @@
+
+    pub fn read_letter_index(&self, key: u32) -> Vec<u8> {
+        let filename = ExtdataArchive::filename_from_key(key);
+        println!("Reading {}...", filename);
```
Do you not want to have this inside `read_file`?

CenTdemeern1 commented on src/extdata.rs:
```diff
@@ -1,167 +1,190 @@
+
+    pub fn write_letter_index(&self, key: u32, data: &[u8]) {
+        let filename = ExtdataArchive::filename_from_key(key);
+        println!("Writing {}...", filename);
```
Ditto for `write_file`

CenTdemeern1 commented on src/phases/rewrite.rs:
```diff
@@ -0,0 +1,218 @@
+    }
+
+    fn do_rewrite(extdata: ExtdataArchive, mapping: OldToNewPIDMapping) {
+        println!("Reading manage.bin...");
```
I'm not gonna write ditto for everything but if you move the other debug print into `read_file` you should remove this and all the other "manual prints"

CenTdemeern1 committed:
style: Copy here instead of dereferencing because it's free

Mainly just looks a bit better, it probably gets optimized into oblivion anyway

CenTdemeern1 committed:
fix: compare cstrings

This is stable since rust 1.90

CenTdemeern1 commented on src/phases/rewrite.rs:
```diff
@@ -0,0 +1,217 @@
+                        .unwrap();
+                let common_block = match letter.iter_mut().find(|k| k.name == c"COMMON1") {
+                    Some(k) => k,
+                    None => continue,
```
Out of curiosity, is this expected to be able to happen?

silver-volt4 replied:
Nope. I'm doing this just to be safe. Perhaps it should be logged though

silver-volt4 replied:
libdoodle doesn't expose the length of COMMON0, perhaps that could be done, but it's out of scope for this PR tbh

CenTdemeern1 replied:
Do you still want to do this?

silver-volt4 replied:
done

CenTdemeern1 commented on src/phases/rewrite.rs:
```diff
@@ -0,0 +1,217 @@
+                extdata.write_letter_index(letter_key, &out);
+            }
+
+            cursor.set_position(pos + 0x80);
```
This feels like a magic relative byte offset. Is this the stride of data in the `COMINF0` block? Should this be a constant?

silver-volt4 replied:
It's len(COMINF0) + 0x40. The 0x40 is specific to COMINF0

silver-volt4 replied:
libdoodle doesn't expose the length of COMMON1, perhaps that could be done, but it's out of scope for this PR tbh

CenTdemeern1 commented on src/phases/rewrite.rs:
```diff
@@ -0,0 +1,217 @@
+        );
+
+        self.gui.end_frame();
+    }
+
+    pub fn paint_progess_page(&self) {
+        self.gui.begin_frame();
+        self.gui.text(
+            &self.progress,
+            TOP_SCREEN_WIDTH / 2.0,
+            TOP_SCREEN_HEIGHT / 2.0 - 40.0,
+            C2D_AlignCenter,
+            0.7,
+        );
+        self.gui.text(
+            &self.progress_observe_bottom,
+            TOP_SCREEN_WIDTH / 2.0,
+            TOP_SCREEN_HEIGHT / 2.0 + 40.0,
+            C2D_AlignCenter,
+            0.5,
+        );
+        self.gui.end_frame();
+    }
+
+    fn paint_done_page(&self) {
+        self.gui.begin_frame();
+        self.gui.header(&self.header_text_finished);
+
+        self.gui.text(&self.finished_line, 10.0, 40.0, 0, 0.5);
+
+        self.gui.text(
+            &self.exit_a,
+            TOP_SCREEN_WIDTH / 2.0,
+            100.0,
+            C2D_AlignCenter,
+            0.7,
+        );
+
+        self.gui.end_frame();
+    }
```
There's a lot of repeated magic numbers here

silver-volt4 replied:
Yeah, they're manually positioned text elements. Not sure how else to handle this though? In theory the UI element positioning could be pre-calculated, but using a variable for the Y coordinate shouldn't be too bad

CenTdemeern1 replied:
My idea was to calculate these at compile time by just starting at a specific offset and adding offsets for the different text elements depending on their size
So the first element is at some starting position
The first element's size is "small" so the next element will be offset down 10 pixels or something
And then that element is medium so the next element will be offset down 20 pixels or something
Rinse and repeat

silver-volt4 replied:
Yeah I'd like to do that too, but I have no idea how tbh

CenTdemeern1 replied:
This is what I've had in mind:
```rust
enum TextSize {
    Small,
    Medium,
    Large
}
use TextSize::*;

enum TextColor {
    White,
    Red,
    Blue,
    Green
}
use TextColor::*;

struct Text<'a> {
    content: &'a str,
    size: TextSize,
    color: TextColor,
}

const EXAMPLE_SCREEN: [Text; 3] = [
    Text {
        content: "Something",
        size: Large,
        color: White,
    },
    Text {
        content: "Something else",
        size: Small,
        color: Red,
    },
    Text {
        content: "Something 3",
        size: Medium,
        color: Blue,
    }
];

/// Start at 100 pixels on the screen
const BASE_OFFSET: usize = 100;

const TEXT_MARGINS: usize = 10;

const OFFSETS: [usize; 3] = {
    let mut offsets = [0; 3]; // Matches `EXAMPLE_SCREEN`'s length
    let mut offset = BASE_OFFSET;
    let mut index = 0;
    // Can't use iterators (including for loops) in const contexts (yet) :/
    loop {
        // Probably not real values, just an example
        let (baseline_offset, full_text_height) = match EXAMPLE_SCREEN[index].size {
            Small => (12, 15),
            Medium => (25, 30),
            Large => (50, 60),
        };
        offsets[index] = offset + baseline_offset;
        offset += full_text_height + TEXT_MARGINS;

        index += 1;
        if index >= EXAMPLE_SCREEN.len() {
            break offsets;
        }
    }
};
```
https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=5107f7c1d90b934c8b6d9e00fda60e56

silver-volt4 replied:
Will be done in another PR after a discussion on Discord

CenTdemeern1 commented on src/phases/reading.rs:
```diff
@@ -0,0 +1,345 @@
+
+        cursor.set_position(pos + 0x80);
+    }
+    println!("done.");
```
This looks very similar to `rewrite.rs`

CenTdemeern1 commented on src/phases/reading.rs:
```diff
@@ -0,0 +1,345 @@
+        {
+            doodles.insert(*row.0, mii);
+        }
+    });
```
Since we're dropping `unknown_pids` after this anyway, might as well consume it in this operation and take ownership of its contents by using `.into_iter()`:
```suggestion
    unknown_pids.into_iter().for_each(|(pid, key)| {
        if let Some(mii) = Letter::new_from_bpk1_bytes(&extdata.read_letter_index(key))
            .unwrap()
            .sender_mii
        {
            doodles.insert(pid, mii);
        }
    });
```

CenTdemeern1 commented on src/phases/reading.rs:
```diff
@@ -0,0 +1,345 @@
+            println!();
+            println!("If you believe this is in error, please let us know!");
+            println!();
+            println!("Press (A) to exit.");
```
You wanted me to remind you that this should have a screen

CenTdemeern1 commented on src/phases/reading.rs:
```diff
@@ -0,0 +1,345 @@
+            println!();
+            println!("If you believe this is in error, please let us know!");
+            println!();
+            println!("Press (A) to exit.");
```
Ditto

CenTdemeern1 commented on src/phases/mapping.rs:
```diff
@@ -0,0 +1,445 @@
+            let mac = doodler.1.creator_mac_address;
+            for friend in &read.friends {
+                if friend.1.creator_mac_address == mac {
+                    mapping.insert(*doodler.0, *friend.0);
```
```suggestion
        for (&doodler_pid, doodler_mii) in &read.doodles {
            for (&friend_pid, friend_mii) in &read.friends {
                if friend_mii.creator_mac_address == doodler_mii.creator_mac_address {
                    mapping.insert(doodler_pid, friend_pid);
```
Giving these values names and copying the PIDs is maybe more readable

CenTdemeern1 commented on src/phases/mapping.rs:
```diff
@@ -0,0 +1,445 @@
+            DIALOG_PADDING + 10.0,
+            DIALOG_PADDING + 25.0,
+            0,
+            0.5,
+        );
+        self.gui.text(
+            &self.explanation_line3,
+            DIALOG_PADDING + 10.0,
+            DIALOG_PADDING + 50.0,
+            0,
+            0.5,
+        );
+        self.gui.text(
+            &self.explanation_line4,
+            DIALOG_PADDING + 10.0,
+            DIALOG_PADDING + 65.0,
+            0,
+            0.5,
+        );
+        self.gui.text(
+            &self.explanation_line5,
+            DIALOG_PADDING + 10.0,
+            DIALOG_PADDING + 90.0,
+            0,
+            0.5,
+        );
+        self.gui.text(
+            &self.explanation_line6,
+            DIALOG_PADDING + 10.0,
+            DIALOG_PADDING + 105.0,
+            0,
+            0.5,
+        );
+        self.gui.text(
+            &self.press_a_continue,
+            DIALOG_PADDING + 10.0,
+            DIALOG_PADDING + 130.0,
+            0,
+            0.7,
+        );
```
This feels too copy-paste-y. You could probably put these `C2D_Text`s in an array? The vertical positions and scale of the text also looks like a pattern but if you don't want to do that you could also `.zip` it with the vertical line position and scale

CenTdemeern1 replied:
Will be dealt with through #2

CenTdemeern1 commented on src/phases/mapping.rs:
A lot of code in this file is assuming hashmaps are ordered, which they aren't.

CenTdemeern1 commented on src/gui/text.rs:
This file should have some comments I wrote a while ago, but a GitHub bug seems to have made them disappear. If this file doesn't have any review comments let me know and I'll copy my old review comments

CenTdemeern1 commented on src/gui/scrollable_view.rs:
This file should have some comments I wrote a while ago, but a GitHub bug seems to have made them disappear. If this file doesn't have any review comments let me know and I'll copy my old review comments

CenTdemeern1 commented on src/gui/mod.rs:
This file should have some comments I wrote a while ago, but a GitHub bug seems to have made them disappear. If this file doesn't have any review comments let me know and I'll copy my old review comments

CenTdemeern1 reviewed (changes requested):
That's a big review
Sorry for taking so long

silver-volt4 committed:
chore: resolve some simple comments

silver-volt4 committed:
feat: use new color constants

silver-volt4 committed:
feat: remove unused dependency

silver-volt4 committed:
fix: rewrite some functions to be more rust-y

silver-volt4 committed:
fix: use vec-based containers for some data

silver-volt4 committed:
feat: add two new screens for rare situations

silver-volt4 committed:
refactor: less repetition in read/rewrite code

silver-volt4 committed:
chore: apply clippy suggestions and run fmt

CenTdemeern1 commented on src/phases/reading.rs:
```diff
@@ -41,45 +40,33 @@ fn friendly_read_data(extdata: &ExtdataArchive) -> (MiiMap, MiiMap) {
             .sender_mii
         {
-            doodles.insert(*row.0, mii);
+            doodles.push((pid, mii));
```
You can use filter_map and collect instead of for_each and push to save on reallocations
(I know you're not using one anymore, [but this would've also worked while the miimap was still a hashmap!](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#impl-FromIterator%3C(K,+V)%3E-for-HashMap%3CK,+V,+S%3E))

CenTdemeern1 reviewed (changes requested):
Changes since last review look good, i've only got one minor nitpick

silver-volt4 committed:
feat: use iterator and skip over malformed letters

silver-volt4 committed:
feat: log if we skipped a letter because of a missing COMMON1 section

CenTdemeern1 reviewed (approved):
looks good to me!

silver-volt4 commented:
congration
